datatransfer/impl/graphsync/graphsync.go

```diff
@@ -116,10 +135,72 @@ func (impl *graphsyncImpl) InProgressChannels() map[datatransfer.ChannelID]datat
 }
 
 // TODO: implement for https://github.com/filecoin-project/go-data-transfer/issues/14
+// ReceiveRequest takes an incoming request, validates the voucher and processes the message.
 func (receiver *graphsyncReceiver) ReceiveRequest(
 	ctx context.Context,
 	sender peer.ID,
 	incoming message.DataTransferRequest) {
+
+	voucher, err := receiver.validateVoucher(sender, incoming)
+	if err != nil {
+		fmt.Printf(err.Error())
+	}
+	if voucher == nil {
+		fmt.Printf("\nincoming voucher failed to validate: %v\n", incoming)
+	}
+}
+
+// validateVoucher converts a voucher in an incoming message to its appropriate
+// voucher struct, then runs the validator and returns the results.
+// returns error if:
+//   * voucherFromRequest fails
+//   * deserialization of selector fails
+//   * validation fails
+func (receiver *graphsyncReceiver) validateVoucher(sender peer.ID, incoming message.DataTransferRequest) (datatransfer.Voucher, error) {
+
+	vtypStr := incoming.VoucherType()
+	vouch, err := receiver.voucherFromRequest(incoming)
+	if err != nil {
+		return vouch,err
+	}
+
+	var validatorFunc func(peer.ID, datatransfer.Voucher, cid.Cid, ipld.Node) error
+	if incoming.IsPull() {
+		validatorFunc = receiver.impl.validatedTypes[vtypStr].validator.ValidatePull
+	} else {
+		validatorFunc = receiver.impl.validatedTypes[vtypStr].validator.ValidatePush
+	}
+
+	stor, err := nodeFromBytes(incoming.Selector())
+	if err != nil { return vouch, err }
+
+	if err = validatorFunc(sender, vouch, incoming.BaseCid(), stor) ; err != nil { return nil, err	}
+
+	return vouch, nil
+}
+
+// voucherFromRequest takes an incoming request and attempts to create a
+// voucher struct from it using the registered validated types.  It returns
+// a deserialized voucher and any error.  It returns error if:
+//    * the voucher type has no validator registered
+//    * the voucher cannot be instantiated via reflection
+//    * request voucher bytes cannot be deserialized via <voucher>.FromBytes()
+func (receiver *graphsyncReceiver) voucherFromRequest(incoming message.DataTransferRequest) (datatransfer.Voucher, error) {
+	vtypStr := incoming.VoucherType()
+
+	validatedType, ok := receiver.impl.validatedTypes[vtypStr]
+	if !ok {
+		return nil, fmt.Errorf("unregistered voucher type %s", vtypStr)
+	}
+	vStructVal := reflect.New(validatedType.voucherType.Elem())
+	voucher, ok := vStructVal.Interface().(datatransfer.Voucher)
+	if !ok || reflect.ValueOf(voucher).IsNil() {
+		return nil, fmt.Errorf("problem instantiating type %s, voucher: %v", vtypStr, voucher)
+	}
+	if err := voucher.FromBytes(incoming.Voucher()); err != nil {
+		return voucher, err
```

Comment:
If this were to error, voucher is most likely an empty voucher of the right type. I don't think that has a lot of value. I would just return nil, err.